File: src/org/codepulsar/pulsar/ByteCodeCompiler.java
package org.codepulsar.pulsar;

import org.codepulsar.analysis.TypeChecker;
import org.codepulsar.analysis.Validator;
import org.codepulsar.ast.Expression;
import org.codepulsar.ast.Statement;
import org.codepulsar.ast.expressions.*;
import org.codepulsar.ast.statements.*;
import org.codepulsar.lang.*;
import org.codepulsar.lang.variables.FunctionVariable;
import org.codepulsar.lang.variables.GlobalVariable;
import org.codepulsar.lang.variables.LocalVariable;
import org.codepulsar.primitives.Primitive;
import org.codepulsar.primitives.PrimitiveType;
import org.codepulsar.primitives.types.*;
import org.codepulsar.util.ASTPrinter;

import java.util.ArrayList;

import static org.codepulsar.lang.ByteCode.*;

public class ByteCodeCompiler implements Expression.Visitor<Instruction>, Statement.Visitor<Void> {
    // Input Data
    private final String sourceCode;
    private Statement program;

    // Data To Help In Compiling To ByteCode
    private FunctionVariable functions;
    private GlobalVariable globals;
    private LocalVariable locals;
    private final ArrayList<Primitive> values; // Constant Values To Be Stored

    // Output Data
    private ArrayList<Instruction> currentChunk;
    private CompilerError errors;
    private CompilerError staticErrors;

    public ByteCodeCompiler(String sourceCode) {
        this.sourceCode = sourceCode;

        this.values = new ArrayList<>();

        this.currentChunk = new ArrayList<>();
    }

    public void compileByteCode() {
        Parser ast = new Parser(this.sourceCode);
        this.program = ast.parse();

        this.functions = ast.getFunctions();
        this.globals = ast.getGlobals();
        this.locals = ast.getLocals();

        this.errors = ast.getErrors();
        if (this.errors.hasError()) return;

        ASTPrinter astPrinter = new ASTPrinter();
        astPrinter.print(this.program);

        TypeChecker analyzer = new TypeChecker(this.program, this.functions, this.globals, this.locals);
        Validator validator = new Validator(this.program, this.functions, this.globals, this.locals);

        validator.validate();
        this.staticErrors = validator.getErrors();
        if (this.staticErrors.hasError()) return;

        analyzer.check();
        this.staticErrors = analyzer.getErrors();
        if (this.staticErrors.hasError()) return;

        compile();
        return;
    }

    private void compile() {
        if (this.program == null) {
            return;
        }

        this.program.accept(this);
    }

    public Void visitVariableStatement(Variable statement) {
        statement.getInitializer().accept(this);
        ByteCode type;

        if (statement.isGlobal()) {
            type = OP_NEW_GLOBAL;
        } else {
            type = OP_NEW_LOCAL;
        }

        makeOpCode(type, statement.getName(), statement.getLine());
        return null;
    }

    public Void visitBlockStatement(Block statement) {
        for (Statement stmt: statement.getStatements()) {
            stmt.accept(this);
        }

        return null;
    }

    public Void visitEndScopeStatement(EndScope statement) {
        for (int i = 0; i < statement.getLocalsToDelete(); i++) {
            makeOpCode(OP_POP, statement.getLine());
        }

        return null;
    }

    public Void visitWhileStatement(While statement) {
        int start = this.currentChunk.size();
        statement.getCondition().accept(this);

        int offset = makeJump(OP_JUMP_IF_FALSE, statement.getLine());
        makeOpCode(OP_POP, statement.getLine());

        statement.getStatements().accept(this);
        makeOpCode(OP_JUMP, start, statement.getStatements().getLine());

        fixJump(offset, OP_JUMP_IF_FALSE);
        makeOpCode(OP_POP, statement.getStatements().getLine());

        return null;
    }

    public Void visitIfStatement(If statement) {
        statement.getCondition().accept(this);
        int ifOffset = makeJump(OP_JUMP_IF_FALSE, statement.getLine());
        makeOpCode(OP_POP, statement.getLine());
        statement.getThenBranch().accept(this);

        int elseOffset = makeJump(OP_JUMP, statement.getThenBranch().getLine());
        fixJump(ifOffset, OP_JUMP_IF_FALSE);
        makeOpCode(OP_POP, statement.getThenBranch().getLine());

        if (statement.hasElse()) {
            statement.getElseBranch().accept(this);
        }

        fixJump(elseOffset, OP_JUMP);
        return null;
    }

    public Void visitNoneStatement(NoneStatement statement) {
        return null;
    }

    public Void visitPrintExpression(Print statement) {
        statement.getExpression().accept(this);
        makeOpCode(OP_PRINT, statement.getLine());

        return null;
    }

    public Void visitReturnStatement(Return statement) {
        statement.getExpression().accept(this);
        makeOpCode(OP_RETURN, statement.getLine());

        return null;
    }

    public Void visitExpressionStatement(ExpressionStmt statement) {
        statement.getExpression().accept(this);
        makeOpCode(OP_POP, statement.getLine());

        return null;
    }

    public Void visitFunctionStatement(Function statement) {
        statement.getStatements().accept(this);
        this.functions.setChunk(statement.getName(), this.currentChunk);

        this.currentChunk = new ArrayList<>();
        return null;
    }

    public Instruction visitAssignmentExpression(Assignment expression) {
        expression.getValue().accept(this);
        ByteCode type;

        if (expression.isGlobalAssignment()) {
            type = OP_STORE_GLOBAL;
            return makeOpCode(type, expression.getIdentifier(), expression.getLine());
        } else {
            type = OP_SET_LOCAL;
            return makeOpCode(type, expression.getNumber(), expression.getLine());
        }
    }

    public Instruction visitBinaryExpression(Binary expression) {
        expression.getLeft().accept(this);
        expression.getRight().accept(this);

        ArrayList<ByteCode> operators = identifyBinaryOperator(expression.getOperator());
        for (ByteCode op: operators) {
            makeOpCode(op, expression.getLine());
        }

        return null;
    }

    public Instruction visitCallExpression(Call expression) {
        makeOpCode(OP_LOAD_FUNCTION, expression.getName().getLiteral(), expression.getLine());
        for (Expression expression1: expression.getArguments()) {
            expression1.accept(this);
        }

        makeOpCode(OP_CALL, expression.getArguments().size(), expression.getLine());
        return null;
    }

    public Instruction visitGroupingExpression(Grouping expression) {
        return expression.getExpression().accept(this);
    }

    public Instruction visitLiteralExpression(Literal expression) {
        String value = expression.getValue();
        return makeConstant(value, expression.getLine(), expression.getType());
    }

    public Instruction visitLogicalExpression(Logical expression) {
        expression.getLeft().accept(this);
        int line = expression.getLine();

        ByteCode jumpType = switch (expression.getOperator()) {
            case "&&" -> OP_JUMP_IF_FALSE;
            case "||" -> OP_JUMP_IF_TRUE;
            default -> null; // Unreachable
        };

        int offset = makeJump(jumpType, line);
        makeOpCode(OP_POP, line);

        expression.getRight().accept(this);
        fixJump(offset, jumpType);

        return null;
    }

    public Instruction visitNoneExpression(NoneExpression expression) {
        return null;
    }

    public Instruction visitUnaryExpression(Unary expression) {
        expression.getRight().accept(this);
        ByteCode operator = identifyUnaryOperator(expression.getOperator());
        return makeOpCode(operator, expression.getLine());
    }

    public Instruction visitVariableExpression(VariableAccess expression) {
        ByteCode type;

        if (expression.isGlobalVariable()) {
            type = OP_LOAD_GLOBAL;
            return makeOpCode(type, expression.getName(), expression.getLine());
        } else {
            type = OP_GET_LOCAL;
            return makeOpCode(type, expression.getNumber(), expression.getLine());
        }
    }

    private Instruction makeConstant(String value, int line, PrimitiveType type) {
        Primitive primitiveLiteral = null;

        switch (type) {
            case PR_INTEGER -> primitiveLiteral = new PInteger(Integer.parseInt(value));
            case PR_DOUBLE -> primitiveLiteral = new PDouble(Double.parseDouble(value));
            case PR_CHARACTER -> primitiveLiteral = new PCharacter(value.charAt(1));
            case PR_BOOLEAN -> primitiveLiteral = new PBoolean(Boolean.parseBoolean(value));
            case PR_NULL -> primitiveLiteral = new PNull();
        }

        this.values.add(primitiveLiteral);

        Instruction instruction = new Instruction(OP_CONSTANT, this.values.size() - 1, line);
        this.currentChunk.add(instruction);
        return instruction;
    }

    private Instruction makeOpCode(ByteCode opcode, int line) {
        Instruction instruction = new Instruction(opcode, null, line);
        this.currentChunk.add(instruction);
        return instruction;
    }

    private Instruction makeOpCode(ByteCode opcode, Object operand, int line) {
        Instruction instruction = new Instruction(opcode, operand, line);
        this.currentChunk.add(instruction);
        return instruction;
    }

    private int makeJump(ByteCode opcode, int line) {
        int size = this.currentChunk.size();
        makeOpCode(opcode, line);

        return size;
    }

    private void fixJump(int offset, ByteCode opcode) {
        Instruction oldJump = this.currentChunk.get(offset);
        int line = oldJump.getLine();
        Instruction jumpOpCode = new Instruction(opcode, this.currentChunk.size(), line);
        this.currentChunk.set(offset, jumpOpCode);
    }

    private ByteCode identifyUnaryOperator(String operator) {
        switch (operator) {
            case "-" -> { return OP_NEGATE; }
            case "!" -> { return OP_NOT; }

            // Unreachable
            default -> { return null; }
        }
    }

    private ArrayList<ByteCode> identifyBinaryOperator(String operator) {
        ArrayList<ByteCode> codes = new ArrayList<>();
        switch (operator) {
            case "+" -> codes.add(OP_ADD);
            case "-" -> codes.add(OP_SUBTRACT);
            case "*" -> codes.add(OP_MULTIPLY);
            case "/" -> codes.add(OP_DIVIDE);
            case "%" -> codes.add(OP_MODULO);

            case ">" -> codes.add(OP_COMPARE_GREATER);
            case ">=" -> {
                codes.add(OP_COMPARE_LESSER);
                codes.add(OP_NOT);
            }

            case "<" -> codes.add(OP_COMPARE_LESSER);
            case "<=" -> {
                codes.add(OP_COMPARE_GREATER);
                codes.add(OP_NOT);
            }

            case "==" -> codes.add(OP_COMPARE_EQUAL);
            case "!=" -> {
                codes.add(OP_COMPARE_EQUAL);
                codes.add(OP_NOT);
            }
        }

        return codes;
    }

    public ArrayList<Primitive> getValues() {
        return this.values;
    }

    public FunctionVariable getFunctions() {
        return this.functions;
    }

    public GlobalVariable getGlobals() {
        return this.globals;
    }

    public LocalVariable getLocals() {
        return this.locals;
    }

    public CompilerError getErrors() {
        return this.errors;
    }

    public CompilerError getStaticErrors() {
        return this.staticErrors;
    }
}

File: src/codepulsar/pulsar/ByteCodeCompiler.cpp
#include "ByteCodeCompiler.h"


Pulsar::ByteCodeCompiler::ByteCodeCompiler(std::string sourceCode) {
    this->sourceCode = sourceCode;

    this->values = std::vector<Primitive*>();
    this->currentChunk = std::vector<Instruction>();
}

std::vector<Pulsar::Instruction> Pulsar::ByteCodeCompiler::compileByteCode() {
    Pulsar::Parser parser = Parser(this->sourceCode);
    this->program = parser.parse();
    this->symbolTable = parser.getSymbolTable();

    this->errors = parser.getErrors();
    if (this->errors->hasError()) return this->currentChunk;

    ASTPrinter rawASTPrinter = ASTPrinter();
    rawASTPrinter.print(this->program, "Raw");

    /* Analyses And Optimizations */ {
        VariableValidator varValidator = VariableValidator(this->program, this->symbolTable, this->errors);
        varValidator.validate();
        if (this->errors->hasError()) return this->currentChunk;

        Validator validator = Validator(this->program, this->symbolTable, this->errors);
        validator.validate();
        if (this->errors->hasError()) return this->currentChunk;

        TypeChecker checker = TypeChecker(this->program, this->symbolTable, this->errors);
        checker.check();
        if (this->errors->hasError()) return this->currentChunk;

        Optimizer optimizer = Optimizer(this->program);
        this->program = optimizer.optimize();
    }

    ASTPrinter optimizedASTPrinter = ASTPrinter();
    optimizedASTPrinter.print(this->program, "Optimized");

    compile();
    return this->symbolTable->getFunctions().find("main")->second.getChunk();
}

void Pulsar::ByteCodeCompiler::compile() {
    if (this->program == nullptr) return;
    this->program->accept(*this);
}

std::any Pulsar::ByteCodeCompiler::visitAssignmentExpression(Assignment* expression) {
    expression->getValue()->accept(*this);

    if (expression->isGlobalAssignment()) {
        return makeOpCode(OP_STORE_GLOBAL, expression->getIdentifier(), expression->getLine());
    } else {
        return makeOpCode(OP_SET_LOCAL, expression->getLocalID(), expression->getLine());
    }
}

std::any Pulsar::ByteCodeCompiler::visitBinaryExpression(Binary* expression) {
    expression->getLeft()->accept(*this);
    expression->getRight()->accept(*this);

    for (ByteCode op: identifyBinaryOperator(expression->getOperator())) {
        makeOpCode(op, expression->getLine());
    }

    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitCallExpression(Call* expression) {
    makeOpCode(OP_LOAD_FUNCTION, expression->getName().literal, expression->getLine());
    for (Expression* expr: *expression->getArguments()) {
        expr->accept(*this);
    }

    makeOpCode(OP_CALL, expression->getName().literal, expression->getLine());
    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitGroupingExpression(Grouping* expression) {
    return expression->getExpression()->accept(*this);
}

std::any Pulsar::ByteCodeCompiler::visitLiteralExpression(Literal* expression) {
    std::string value = expression->getValue();
    return makeConstant(value, expression->getType(), expression->getLine());
}

std::any Pulsar::ByteCodeCompiler::visitLogicalExpression(Logical* expression) {
    expression->getLeft()->accept(*this);
    int line = expression->getLine();
    ByteCode jumpType;

    if (expression->getOperator() == "&&") { jumpType = OP_JUMP_IF_FALSE; }
    else if (expression->getOperator() == "||") { jumpType = OP_JUMP_IF_TRUE; }

    int offset = makeJump(jumpType, line);
    makeOpCode(OP_POP, line);

    expression->getRight()->accept(*this);
    fixJump(jumpType, offset);

    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitUnaryExpression(Unary* expression) {
    expression->getExpression()->accept(*this);
    ByteCode op = identifyUnaryOperator(expression->getOperator());
    return makeOpCode(op, expression->getLine());
}

std::any Pulsar::ByteCodeCompiler::visitVariableExpression(VariableExpr* expression) {
    if (expression->isGlobalVariable()) {
        return makeOpCode(OP_LOAD_GLOBAL, expression->getName(), expression->getLine());
    } else {
        return makeOpCode(OP_GET_LOCAL, expression->getLocalID(), expression->getLine());
    }
}

std::any Pulsar::ByteCodeCompiler::visitBlockStatement(Block* statement) {
    for (Statement* stmt: *statement->getStatements()) {
        stmt->accept(*this);
    }

    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitExpressionStatement(ExpressionStmt* statement) {
    statement->getExpression()->accept(*this);
    makeOpCode(OP_POP, statement->getLine());
    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitFunctionStatement(Function* statement) {
    statement->getStatements()->accept(*this);
    this->symbolTable->setChunk(statement->getName(), this->currentChunk);
    this->currentChunk = std::vector<Instruction>();
    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitIfStatement(If* statement) {
    statement->getCondition()->accept(*this);
    int ifOffset = makeJump(OP_JUMP_IF_FALSE, statement->getLine());
    makeOpCode(OP_POP, statement->getLine());
    statement->getThenBranch()->accept(*this);

    int elseOffset = makeJump(OP_JUMP, statement->getThenBranch()->getLine());
    fixJump(OP_JUMP_IF_FALSE, ifOffset);
    makeOpCode(OP_POP, statement->getThenBranch()->getLine());

    if (statement->hasElse()) {
        statement->getElseBranch()->accept(*this);
    }

    fixJump(OP_JUMP, elseOffset);
    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitNoneStatement(NoneStmt* statement) {
    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitPrintStatement(Print* statement) {
    statement->getExpression()->accept(*this);
    makeOpCode(OP_PRINT, statement->getLine());
    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitProgramStatement(Program* statement) {
    for (Statement* stmt: *statement->getStatements()) {
        stmt->accept(*this);
    }

    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitReturnStatement(Return* statement) {
    if (statement->hasValue()) {
        statement->getValue()->accept(*this);
    }

    makeOpCode(OP_RETURN, statement->getLine());
    return nullptr;
}

std::any Pulsar::ByteCodeCompiler::visitVariableStatement(VariableDecl* statement) {
    statement->getInitializer()->accept(*this);

    if (statement->isGlobalVariable()) {
        return makeOpCode(OP_NEW_GLOBAL, statement->getName(), statement->getLine());
    } else {
        return makeOpCode(OP_NEW_LOCAL, statement->getName(), statement->getLine());
    }
}

std::any Pulsar::ByteCodeCompiler::visitWhileStatement(While* statement) {
    int start = this->currentChunk.size();
    statement->getCondition()->accept(*this);

    int offset = makeJump(OP_JUMP_IF_FALSE, statement->getLine());
    makeOpCode(OP_POP, statement->getLine());

    statement->getStatements()->accept(*this);
    makeOpCode(OP_JUMP, start, statement->getStatements()->getLine());

    fixJump(OP_JUMP_IF_FALSE, offset);
    makeOpCode(OP_POP, statement->getStatements()->getLine());

    return nullptr;
}

Pulsar::Instruction Pulsar::ByteCodeCompiler::makeConstant(std::string value, PrimitiveType type, int line) {
    if (type == PR_INTEGER) {
        this->values.emplace_back(new PInteger(std::stoi(value)));
    } else if (type == PR_DOUBLE) {
        this->values.emplace_back(new PDouble(std::stod(value)));
    } if (type == PR_CHARACTER) {
        this->values.emplace_back(new PCharacter(value[1]));
    } if (type == PR_BOOLEAN) {
        this->values.emplace_back(new PBoolean(value == "true"));
    }

    return makeOpCode(OP_CONSTANT, (int) (this->values.size() - 1), line);
}

Pulsar::Instruction Pulsar::ByteCodeCompiler::makeOpCode(ByteCode opcode, int line) {
    return makeOpCode(opcode, nullptr, line);
}

Pulsar::Instruction Pulsar::ByteCodeCompiler::makeOpCode(ByteCode opcode, std::any operand, int line) {
    Instruction instruction = Instruction(opcode, operand, line);
    this->currentChunk.push_back(instruction);
    return instruction;
}

int Pulsar::ByteCodeCompiler::makeJump(ByteCode opcode, int line) {
    int size = this->currentChunk.size();
    makeOpCode(opcode, line);
    return size;
}

void Pulsar::ByteCodeCompiler::fixJump(ByteCode opcode, int offset) {
    Instruction oldJump = this->currentChunk[offset];
    int line = oldJump.getLine();
    Instruction jumpOpCode = Instruction(opcode, (int) this->currentChunk.size(), line);
    this->currentChunk.at(offset) = jumpOpCode;
}

Pulsar::ByteCode Pulsar::ByteCodeCompiler::identifyUnaryOperator(std::string oper) {
    if (oper == "-") { return OP_NEGATE; }
    else if (oper == "!") { return OP_NOT; }
}

std::vector<Pulsar::ByteCode> Pulsar::ByteCodeCompiler::identifyBinaryOperator(std::string oper) {
    std::vector<ByteCode> codes = std::vector<ByteCode>();

    if (oper == "+") { codes.push_back(OP_ADD); }
    else if (oper == "-") { codes.push_back(OP_SUBTRACT); }
    else if (oper == "*") { codes.push_back(OP_MULTIPLY); }
    else if (oper == "/") { codes.push_back(OP_DIVIDE); }
    else if (oper == "%") { codes.push_back(OP_MODULO); }

    else if (oper == ">") { codes.push_back(OP_COMPARE_GREATER); }
    else if (oper == ">=") {
        codes.push_back(OP_COMPARE_LESSER);
        codes.push_back(OP_NOT);
    }

    else if (oper == "<") { codes.push_back(OP_COMPARE_LESSER); }
    else if (oper == "<=") {
        codes.push_back(OP_COMPARE_GREATER);
        codes.push_back(OP_NOT);
    }

    else if (oper == "==") { codes.push_back(OP_COMPARE_EQUAL); }
    else if (oper == "!=") {
        codes.push_back(OP_COMPARE_EQUAL);
        codes.push_back(OP_NOT);
    }

    return codes;
}

Pulsar::SymbolTable* Pulsar::ByteCodeCompiler::getSymbolTable() {
    return this->symbolTable;
}

std::vector<Pulsar::Primitive*> Pulsar::ByteCodeCompiler::getValues() {
    return this->values;
}

Pulsar::CompilerError* Pulsar::ByteCodeCompiler::getErrors() {
    return this->errors;
}
